src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,143 @@
+                                DataVersion aDataVersion);
+
+    /**
+     * FinishAttribute encodes the "footer" part of an attribute report (it closes the containers opened in PrepareAttribute)
+     */
+    CHIP_ERROR FinishAttribute();
+
+    template <typename... Ts>
+    CHIP_ERROR Encode(Ts... aArgs)
+    {
+        return DataModel::Encode(*mAttributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)),
+                                 std::forward<Ts>(aArgs)...);
+    }
+
+private:
+    AttributeReportIB::Builder mAttributeReportIBBuilder;
+    AttributeDataIB::Builder mAttributeDataIBBuilder;
+};
+
+/**
+ * The AttributeValueEncoder is a helper class for filling report payloads into AttributeReportIBs.
+ * The attribute value encoder can be initialized with a AttributeEncodeState for saving and recovering its state between encode
+ * sessions (chunkings).
+ *
+ * When Encode returns recoverable errors (e.g. CHIP_ERROR_NO_MEMORY) the state can be used to initialize the AttributeValueEncoder
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
+{
+public:
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
+    };
+
+    /**
+     * AttributeReportBuilder offleads some code in AttributeValueEncoder::Encode to keep its content small.
+     */
+    AttributeValueEncoder(AttributeReportIBs::Builder & aAttributeReportIBsBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, DataVersion aDataVersion,
+                          const AttributeEncodeState & aState = AttributeEncodeState()) :
+        mAttributeReportIBsBuilder(aAttributeReportIBsBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mDataVersion(aDataVersion), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+        AttributeReportBuilder builder;
+
+        ReturnErrorOnFailure(builder.PrepareAttribute(
+            mAttributeReportIBsBuilder,
+            ConcreteDataAttributePath(mPath.mEndpointId, mPath.mClusterId, mPath.mAttributeId,
```

Comment:
So here's a question.  Can we just store a ConcreteDataAttributePath as a member, and then update its state as needed (ideally in not-per-type code), instead of recomputing it here in per-type code?

Followup is OK for this part.